lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -140,6 +140,37 @@ def default_insert_value(column)
         private :default_insert_value
 
         def build_insert_sql(insert) # :nodoc:
+          if insert.skip_duplicates?
+            # Do we have a unique_by index? Use index columns
+            conflict_columns = if (unique_by = insert.send(:insert_all).unique_by)
+              [unique_by.columns]
+            else
+              # Compare against every unique constraint (primary key included).
+              # Discard constraints that are not fully included on insert.keys. Prevents invalid queries.
+              # Example: ignore unique index for columns ["name"] if insert keys is ["description"]
+              insert_all = insert.send(:insert_all)
+
+              (insert_all.send(:unique_indexes).map(&:columns) + [insert_all.primary_keys]).select do |columns|
+                columns.to_set.subset?(insert.keys)
+              end
+            end
+            includes_primary_key = (insert.send(:insert_all).primary_keys.to_set & insert.keys).present?
+
+            sql = +""
+            sql << "SET IDENTITY_INSERT #{insert.model.quoted_table_name} ON;" if includes_primary_key
+            sql << "MERGE INTO #{insert.model.quoted_table_name} WITH (UPDLOCK, HOLDLOCK) AS target"
+            sql << " USING (SELECT DISTINCT * FROM (#{insert.values_list}) AS t1 (#{insert.send(:columns_list)})) AS source"
```

Comment:
None of the standard adapters uniquify insert rows. I don't think we need to do it here either.

Reply:
uniquify was the solution to this test/scenario
```
Book.insert_all [
  { author_id: 8, name: "Refactoring" },
  { author_id: 8, name: "Refactoring" }
]
```
I don't like it but it works so I move on until all test pass.

Then I found the failing test (still stuck with it)
```
Book.insert_all [
  { id: 200, author_id: 8, name: "Refactoring" },
  { id: 201, author_id: 8, name: "Refactoring" }
]
```
My impression is that the solution to this will let me remove the uniquify.